source/robot_model/src/Model.cpp

```diff
@@ -260,9 +263,112 @@ std::vector<state_representation::CartesianPose> Model::forward_geometry(const s
   return this->forward_geometry(joint_state, frame_ids);
 }
 
-state_representation::JointPositions Model::inverse_geometry(const state_representation::CartesianState&) const {
-  // TODO
-  return state_representation::JointPositions();
+Eigen::MatrixXd Model::cwln_weighted_matrix(const state_representation::JointPositions& joint_positions,
+                                            const double margin) {
+  Eigen::MatrixXd W_b = Eigen::MatrixXd::Identity(this->robot_model_.nq, this->robot_model_.nq);
+  for (int n = 0; n < this->robot_model_.nq; ++n) {
+    double d = 1;
+    W_b(n, n) = 1;
+    if (joint_positions.get_positions()[n] < this->robot_model_.lowerPositionLimit[n] + margin) {
+      if (joint_positions.get_positions()[n] < this->robot_model_.lowerPositionLimit[n]) {
+        W_b(n, n) = 0;
+      } else {
+        d = (this->robot_model_.lowerPositionLimit[n] + margin - joint_positions.get_positions()[n]) / margin;
+        W_b(n, n) = -2 * d * d * d + 3 * d * d;
+      }
+    } else if (this->robot_model_.upperPositionLimit[n] - margin < joint_positions.get_positions()[n]) {
+      if (this->robot_model_.upperPositionLimit[n] < joint_positions.get_positions()[n]) {
+        W_b(n, n) = 0;
+      } else {
+        d = (joint_positions.get_positions()[n] - (this->robot_model_.upperPositionLimit[n] - margin)) / margin;
+        W_b(n, n) = -2 * d * d * d + 3 * d * d;
+      }
+    }
+  }
+  return W_b;
+}
+
+Eigen::VectorXd Model::cwln_repulsive_potential_field(const state_representation::JointPositions& joint_positions,
+                                                      double margin) {
+  Eigen::VectorXd Psi(this->robot_model_.nq);
+  Eigen::VectorXd q = joint_positions.get_positions();
+  for (int i = 0; i < this->robot_model_.nq; ++i) {
+    Psi[i] = 0;
+    if (q[i] < this->robot_model_.lowerPositionLimit[i] + margin) {
+      Psi[i] =
```

Comment:
can we format that differently by adding a linebreak in the middle of the operation?